src/dodal/devices/zocalo/zocalo_results.py

```diff
@@ -0,0 +1,224 @@
+        for i in range(3)
+    ]
+
+
+class ZocaloResults(StandardReadable, Triggerable):
+    """An ophyd device which can wait for results from a Zocalo job"""
+
+    def __init__(
+        self,
+        name: str = "zocalo_results",
+        zocalo_environment: str = "dev_artemis",
+        channel: str = "xrc.i03",
+        sort_key: str = DEFAULT_SORT_KEY.value,
+        timeout_s: float = DEFAULT_TIMEOUT,
+        prefix: str = "",
+    ) -> None:
+        self.zocalo_environment = zocalo_environment
+        self.sort_key = SortKeys[sort_key]
+        self.channel = channel
+        self.timeout_s = timeout_s
+        self._prefix = prefix
+
+        self._subscription_run: bool = False
+        self._raw_results_received: asyncio.Queue = asyncio.Queue()
+
+        self.results = create_soft_signal_r(list[XrcResult], "results", self.name)
+        self.centres_of_mass = create_soft_signal_r(
+            NDArray[np.uint64], "centres_of_mass", self.name
+        )
+        self.bbox_sizes = create_soft_signal_r(
+            NDArray[np.uint64], "bbox_sizes", self.name
+        )
+        self.set_readable_signals(
+            read=[
+                self.results,
+                self.centres_of_mass,
+                self.bbox_sizes,
+            ]
+        )
+        super().__init__(name)
+
+    async def _put_results(self, results: Sequence[XrcResult]):
+        await self.results._backend.put(list(results))
+        centres_of_mass = np.array([r["centre_of_mass"] for r in results])
+        bbox_sizes = np.array([bbox_size(r) for r in results])
+        await self.centres_of_mass._backend.put(centres_of_mass)
+        await self.bbox_sizes._backend.put(bbox_sizes)
+
+    @AsyncStatus.wrap
+    async def trigger(self):
+        """Returns an AsyncStatus waiting for results to be received from Zocalo."""
+        LOGGER.info("Zocalo trigger called")
+        if not self._subscription_run:
+            LOGGER.info("subscription not initialised, subscribing to queue")
+            self._subscribe_to_results()
+            self._subscription_run = True
+
+        try:
+            LOGGER.info("waiting for results in queue")
+
+            async def _get_results():
+                raw_results = await self._raw_results_received.get()
+                LOGGER.info(f"Zocalo: found {len(raw_results)} crystals.")
+                # Sort from strongest to weakest in case of multiple crystals
+                await self._put_results(
+                    sorted(
+                        raw_results, key=lambda d: d[self.sort_key.value], reverse=True
+                    )
+                )
+
+            task = asyncio.create_task(_get_results())
+            try:
+                await asyncio.wait_for(asyncio.shield(task), self.timeout_s / 2)
+            except TimeoutError:
+                LOGGER.warning("Waited half of timeout for zocalo results - retrying")
+                await asyncio.wait_for(task, self.timeout_s / 2)
```

Comment:
Should: Do we need this? It seems only really useful for testing?

Reply:
I think this is important to avoid unnecessary crashes if we have a repeat of the issue we saw while testing. sure, maybe in  the long term this shouldn't be the solution, but it's at least necessary for soak testing imo.

Reply:
Ok, can we have some comments here about the issue we saw and so why we're doing a retry. Also, if we think we want a different long term solution can you make an issue for it?